Add card URL bar

FadhlanR opened the pull request:
https://github.com/cardstack/boxel/assets/12637010/8f758988-bcc7-408a-b4b3-01c56107f05e

FadhlanR committed:
Add card URL bar

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -95,12 +102,19 @@ export default class OperatorModeContainer extends Component<Signature> {
     super(owner, args);
 
     this.messageService.register();
+    scheduleOnce('afterRender', this, () =>
```
why are we scheduling this after render? it's an async task does it matter when it's kicked off?

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -493,9 +507,12 @@ export default class OperatorModeContainer extends Component<Signature> {
+  get lastCardInRightMostStack(): CardDef {
+    return (
+      this.allStackItems
+        // @ts-ignore Property 'card' does not exist on type 'StackItem'. - it actually does exist because we filtered for it in the line above
```
it does exists if you merge upstream changes. please remove this ignore. in teh future you should solve this by typing correctly (and casting as necessary) not by ignoring errors

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+        this.onCodePathChange(new URL(this.lastCardInRightMostStack.id));
+        break;
+    }
+
```
please include a default case. it should assert never. this is a good type script practice that will catch issues if we ever add a new submode. an example is here:

https://github.com/cardstack/boxel/blob/549a5b72cc8aa978e4c5ab15fdb602b9a6b54391/packages/runtime-common/loader.ts#L722-L723

jurgenwerk commented:
I I took this for a spin and I noticed the following behaviours which I think can be improved:

1. Looking at the url bar it does not look to me that this is actually a URL:

<img width="487" alt="image" src="https://github.com/cardstack/boxel/assets/273660/550e5d4a-639f-43b0-817f-318994b5b631">

Only when clicking into it, the URL shows up. I find this a bit unintuitive. I know this is the behaviour in Safari but to me personally it would be more clear and simple if we just displayed the whole URL. It would also be much more convenient for inspection and debugging

2. When clicking the URL, the left portion will disappear. This feels jarring because the position of thing that you are editing changes. I think it would be better if it didn't disappear: 
<img width="350" alt="image" src="https://github.com/cardstack/boxel/assets/273660/0efb4972-fb28-429d-b75b-779662445366">

<img width="299" alt="image" src="https://github.com/cardstack/boxel/assets/273660/d19aeea8-87ee-4d12-a1fe-3d26181b8274">

3. Unexpected cursor jumps
<img width="677" alt="image" src="https://github.com/cardstack/boxel/assets/273660/7d30688f-3210-40bd-818d-214d203d557a">

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+  @action
+  onKeyPress(event: KeyboardEvent) {
+    try {
+      if (event.key === 'Enter' || event.keyCode === 13) {
```
Is it necessary to use both conditions? Does only the `event.key === 'Enter'` not suffice?

habdelra replied:
confirm @jurgenwerk's suggestion is the modern way to do this

jurgenwerk commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
   }
 
+  @action onCodePathChange(codePath: URL | null) {
+    this.operatorModeStateService.updateCodePath(codePath);
```
Will this update code path even when it it's invalid/wrong?

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+      data-test-card-url-bar
+      ...attributes
+    >
+      {{#if (not this.isFocused)}}
```
You could also use `{{#unless this.isFocused}}`

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+  <template>
+    <div
+      id='card-url-bar'
+      class={{this.cssStyle}}
```
`class={{this.cssClasses}}` since you're returning classes and not actual css style

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+      </div>
+      {{#if this.showErrorMessage}}
+        <div class='error-message' data-test-card-url-bar-error>
+          <span>{{if this.isInvalid this.errorMessage @errorMessage}}</span>
```
This whole block reads confusingly. There is logic in `showErrorMessage` that operates on `this.errorMessage` and `@errorMessage`, and then there is also logic in the template here that again operates on the same params. I think this can be simplified

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+          const realmPath = new RealmPaths(realmURL);
+          return realmPath.inRealm(codePath);
+        }),
+      );
```
let's turn this into a method that the card-service can provide. There are other cases where there is a need to see if a URL is in a particular realm.

it could be: `CardService.getRealmForURL(url: URL): URL | undefined` where you give it a URL and the method will return the realm that contains the given URL or undefined if it matches no realms

habdelra replied:
actually i just added this for an unrelated feature that also needed this functionality. if you merge upstream changes you'll get this: https://github.com/cardstack/boxel/blob/d0a5999b4ae29c9d9363a91fe421e6a2a0c7350b/packages/host/app/services/card-service.ts#L324-L332

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+          const realmPath = new RealmPaths(realmURL);
+          return realmPath.inRealm(codePath);
+        }),
+      );
```
also why is this async? you are not `await`ing inside the callback. It doesn't look like there is any need for a Promise.all. a normal for loop should work fine here.

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+          return realmPath.inRealm(codePath);
+        }),
+      );
+      if (!isInAnyRealms.includes(true))
```
this is an awkward test--an array of bool values is a weird thing to return, you should just return the realm that the URL is in. probably this is a side effect of using a Promise.all--but you are doing a promise.all around synchronous code which doesn't make sense.

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -590,9 +607,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+
+      this.cardForCardMode = await this.cardService.loadModel(codePath);
+      this.isCodePathInvalid = false;
+      this.codePathErrorMessage = null;
```
generally you'd want to set these values _before_ the async, not after. if you had an error and typed a new URL, the errors should reset as soon as you start typing not after you try to load the card. I haven't gotten that far in my review, but I think that you need an action that the URL text field calls when the value changes. one of the things this action should do is to reset the error state so that as soon as a user starts typing after getting an error, the error state is cleared.

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -493,9 +507,12 @@ export default class OperatorModeContainer extends Component<Signature> {
+    return (
+      this.allStackItems
+        // @ts-ignore Property 'card' does not exist on type 'StackItem'. - it actually does exist because we filtered for it in the line above
+        .reverse()[0].card
```
reverse mutates an array in place. would strongly advise not to use this unless you really wanna mutate this array.

rather, just `this.allStackItems[this.allStackItems.length - 1]`

habdelra commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+  Element: HTMLElement;
+  Args: {
+    url: URL;
+    onURLChange: (url: URL) => void;
```
so it looks like this callback is invoked when the URL changes. you should add the logic to clear the error state in this callback so that as soon as a user starts typing after an error the error message disappears

habdelra commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,190 @@
+      if (event.key === 'Enter' || event.keyCode === 13) {
+        this.args.onURLChange(new URL(this.url));
+        this.isInvalid = false;
+        this.errorMessage = null;
```
this is awkward, you have 2 places that are the source for error messages--aka 2 roots of state for error messages: `this.args.errorMessage` and `this.errorMessage`. both of these are determining the same things--if the URL is valid. you need to resolve this conflict--there should be only one place that is determining if a URL is valid. you then use callbacks (actions) and args to coordinate that. There is a saying in ember: "Data down, actions up" meaning that the parents should set properties on their child components, and child components should bubble up actions (invoke callbacks that the parent sent them) in order to change state in their parents. You need to decide who owns the error message state: the container or the card-url-bar (probably it would make the most sense if the url bar owns its own error state). Please untangle this.

FadhlanR committed:
Simplify lastCardInRightMostStack function

FadhlanR committed:
Add assertNever

FadhlanR committed:
Merge branch 'main' into cs-5838-card-url-bar

FadhlanR committed:
Update based on feedback

FadhlanR committed:
remove unnecessary function and field

jurgenwerk commented:
@FadhlanR can you make this a bit easier to review again? There was a bunch of comments and no replies apart from new code. I suggest adding a summary/description of the notable adjustments, adding screenshots/videos of things that changed, etc. All this helps with making the review process easier. I would also like to get your opinions regarding https://github.com/cardstack/boxel/pull/571#issuecomment-1693432943.

FadhlanR commented:
I have made the necessary updates to the code based on all the provided feedback. Here is the list of changes that should be highlighted:

- [x] Added the `assertNever` call to the `updateSubmode` function.
- [x] Simplified the implementation of the `lastCardInRightMostStack` function and removed any unnecessary comments.
- [x] Utilized the `cardService.getRealmURLFor` function to validate whether the URL corresponds to a recognized realm.
- [x] Revised the error state names in the code URL component. The `isInvalidURL` state reflects an internal error condition, signifying the validity of the user-input string as a URL. Meanwhile, the `notFoundError` argument indicates the occurrence of a "not found" error originating from the parent component.
- [x] The `codePath` is only updated when the URL points to a valid Card URL.
- [x] Adjusted the behavior of code URLs to ensure the omission of the card type's name display and the consistent visibility of realm information, irrespective of the scenario.

Here is a screenshot for your reference:


https://github.com/cardstack/boxel/assets/12637010/9f4827cb-9e79-44bf-930f-f05fcade8858


(Note: Please disregard the sub-mode switcher button, as it appears to be an error from the main branch. I will address this issue in a separate pull request.)

FadhlanR committed:
lint fix

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -602,14 +653,24 @@ export default class OperatorModeContainer extends Component<Signature> {
+          {{#if this.isCodeMode}}
+            <CardURLBar
+              @url={{this.codePath}}
+              @onEnterPressed={{this.loadCardForCodeMode}}
```
good job with unwinding the multiple roots of state you had before

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -602,14 +653,24 @@ export default class OperatorModeContainer extends Component<Signature> {
+              @url={{this.codePath}}
+              @onEnterPressed={{this.loadCardForCodeMode}}
+              @card={{this.cardForCardMode}}
+              @notFoundError={{this.cardNotFoundError}}
```
i would rename this to `codeModeCardError` that makes it more clear what this error is about.

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -92,6 +95,9 @@ export default class OperatorModeContainer extends Component<Signature> {
   @tracked isChatVisible = false;
 
+  @tracked cardForCardMode: CardDef | null = null;
+  @tracked cardNotFoundError: string | null = null;
```
i would rename this to codeModeCardError that makes it more clear what this error is about.

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -92,6 +95,9 @@ export default class OperatorModeContainer extends Component<Signature> {
   @tracked searchSheetTrigger: SearchSheetTrigger | null = null;
   @tracked isChatVisible = false;
 
+  @tracked cardForCardMode: CardDef | null = null;
```
this name is a bit weird. how about `codeModeCard` instead?

habdelra commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,178 @@
+      this.isInvalidURL = true;
+    }
+
+    if (url) this.args.onEnterPressed(url);
```
let's not do `if` statements without `{}` it's too easy to introduce bugs if we decide we want to add more statements for the `if`

habdelra replied:
we will need to reset the error state in the container when a non-enter key is pressed. I think you'll need to pass another callback argument into this component that allows us to reset the error state if the user starts typing

FadhlanR replied:
> we will need to reset the error state in the container when a non-enter key is pressed. I think you'll need to pass another callback argument into this component that allows us to reset the error state if the user starts typing

I did it here https://github.com/cardstack/boxel/pull/571/files#diff-7e7fc64a8985fde5bec07503f08e35be28fdfe2e4622996d55595dbde3f0e7a8R18.

tintinthong commented:
@FadhlanR I was wondering. When you input a card url this function is only reserved for card instances but not modules/gts files? The `.loadModel` is called which essentially will load a card instance

habdelra commented:
Good point. We’ll definitely need to load modules too.

FadhlanR commented:
> @FadhlanR I was wondering. When you input a card url this function is only reserved for card instances but not modules/gts files? The `.loadModel` is called which essentially will load a card instance

What if we do that on another ticket that really use that modules? Maybe we have a ticket to implement the code editor in code mode. I think @habdelra  is working on that https://linear.app/cardstack/issue/CS-5902/show-monaco-editor-with-open-file-in-center-panel.

FadhlanR committed:
renaming fields and arguments

FadhlanR committed:
Add bracket for if

habdelra commented:
@FadhlanR the PR i was picking up was just around showing monaco.  it wasnt specifically for modules. @lukemelia should we create a new ticket for loading modules specifically in code mode? or was the intent for this ticket that it encompase both instances and modules?

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -587,9 +595,54 @@ export default class OperatorModeContainer extends Component<Signature> {
+    this.loadCardForCodeModeTask.perform(codePath);
+  }
+
+  loadCardForCodeModeTask = restartableTask(async (codePath: URL) => {
```
i think ultimately this will have to be refactored for loading a "file" in code mode, not a card--using the file resource that we currently use in our existing code route.

FadhlanR replied:
I see, I think we need to load both the card and the file since the card is used in the code URL bar to show realm info.

lukemelia replied:
Is there another mechanism we can use to find the realm? Perhaps based on just the URL?

lukemelia commented:
> @FadhlanR the PR i was picking up was just around showing monaco. it wasnt specifically for modules. @lukemelia should we create a new ticket for loading modules specifically in code mode? or was the intent for this ticket that it encompase both instances and modules?

This should be able to open any file in the realm that can be shown in the monaco editor.

burieberry reviewed (commented):
I'm wondering if we can utilize the `search-input` boxel component, which has all the validation, focus, event handling features, instead of writing a new component. I did do some refactor to it in my PR, but maybe you and/or I could do a follow-up PR after it's merged to use that component

burieberry commented on packages/boxel-ui/public/assets/icon-globe.svg:
```diff
@@ -0,0 +1,6 @@
+<!-- Uploaded to: SVG Repo, www.svgrepo.com, Generator: SVG Repo Mixer Tools -->
+<svg fill="var(--icon-color, #000)" width="800px" height="800px" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+    <path
+        d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm7.931 9h-2.764a14.67 14.67 0 0 0-1.792-6.243A8.013 8.013 0 0 1 19.931 11zM12.53 4.027c1.035 1.364 2.427 3.78 2.627 6.973H9.03c.139-2.596.994-5.028 2.451-6.974.172-.01.344-.026.519-.026.179 0 .354.016.53.027zm-3.842.7C7.704 6.618 7.136 8.762 7.03 11H4.069a8.013 8.013 0 0 1 4.619-6.273zM4.069 13h2.974c.136 2.379.665  ...
```
It would be good to pass any svg/images through a compressor, such as ImageOptim or other, before uploading them. You will probably need to add the `fill="var(--icon-color, #000)` back in because it might be removed

jurgenwerk replied:
I usually run it through https://svgomg.net/. Just noting that this one works well too

burieberry commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,188 @@
+        display: flex;
+        align-items: center;
+
+        background: var(--boxel-purple-700);
```
it's sufficient to just use `background-color` property when that's the only thing you're changing

burieberry commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,188 @@
+        position: relative;
+      }
+      .focused {
+        outline: 2px solid var(--boxel-teal);
```
it'd be better to use `var(--boxel-highlight)` instead of the `var(--boxel-teal)` to keep all focus/highlight styles consistent

burieberry commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -0,0 +1,188 @@
+
+        color: var(--boxel-light);
+      }
+      .error-message {
+        position: absolute;
+        bottom: calc(calc(var(--boxel-sp-xs) * 2) * -1);
+        color: red;
+      }
```
we already have a set of error styles that come with the BoxelInput component. Why was it necessary to write new ones? Also we have a couple of `var(--boxel-error...)` colors for the error red

burieberry commented:
Another thing I noticed is any dark colored icon will not be very visible on this search bar. We can already see this issue with the black Cardstack logo for the Drafts realm. We should get some design feedback about this

habdelra commented:
> > @FadhlanR the PR i was picking up was just around showing monaco. it wasnt specifically for modules. @lukemelia should we create a new ticket for loading modules specifically in code mode? or was the intent for this ticket that it encompase both instances and modules?
> 
> This should be able to open any file in the realm that can be shown in the monaco editor.

i think this means that we also need to incorporate the node-like resolution that the current code route does. I.e. for modules you should be able to load these either with or without their extensions

FadhlanR committed:
Fetch card resource instead

FadhlanR committed:
Update css and icon based on feedback

FadhlanR commented:
After a discussion with @lukemelia  and @habdelra , we reached a decision to load card modules instead of loading card instances while in code mode. The most recent commit includes the addition of the `openFile` function, designed as a file resource for loading the card modules. This function will automatically update as the `codePath` gets updated. I also have added functions to retrieve realm information using the realm URL. Consequently, having a card in code mode to fetch realm information is no longer a requirement.

tintinthong commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -611,38 +614,60 @@ export default class OperatorModeContainer extends Component<Signature> {
+    const relativePath = this.codePath
+      .toString()
+      .replace(realmURL.toString(), '');
+    console.log(relativePath);
```
Can we use RealmPaths.local for relativePath

tintinthong commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -611,38 +614,60 @@ export default class OperatorModeContainer extends Component<Signature> {
     return (
       this.operatorModeStateService.state.codePath ??
-      new URL(this.lastCardInRightMostStack.id)
+      new URL(this.lastCardInRightMostStack.id + '.json')
```
oh wait why is this default needed?

EDIT: Oh its for the types to check out properly

tintinthong replied:
For one. this.lastCardInRightMostStack types are misleading. It doesn't guarantee the CardDef return. During runtime, theres a case where the right most card doesn't exist


https://github.com/cardstack/boxel/assets/8165111/a761a414-ef72-4bc0-963f-f99ec541c2de

tintinthong replied:
id favor just using index card of this realm in there

tintinthong commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -587,9 +599,75 @@ export default class OperatorModeContainer extends Component<Signature> {
+    this.loadFileError = null;
+  }
+
+  openFile = maybe(this, (context) => {
```
👍 thx for incorporating this

tintinthong reviewed (approved):
👍  minor changes suggested

FadhlanR committed:
Merge branch 'main' into cs-5838-card-url-bar

FadhlanR committed:
address feedback

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -57,6 +62,7 @@ interface Signature {
 export interface OperatorModeState {
```
just a general observation--why does this interface live in this component when it is only ever used in the operator-mode-state-service? probably that is where it should live...

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -586,9 +602,76 @@ export default class OperatorModeContainer extends Component<Signature> {
+  updateCodePath(codePath: URL | null) {
+    this.operatorModeStateService.updateCodePath(codePath);
+    this.fetchCodeModeRealmInfo.perform();
+  }
```
i'm confused by this--it looks like the container component has a codePath state, but also the operatorModeStateService has a codePath state. why do we have to track the code path as state in 2 different places? what happens when they are different?

FadhlanR replied:
Updated here https://github.com/cardstack/boxel/pull/571/commits/0f19726343a7b6185e7960aae0878210facb4b4e

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -601,14 +684,27 @@ export default class OperatorModeContainer extends Component<Signature> {
+              @loadFileError={{this.loadFileError}}
+              @resetLoadFileError={{this.resetLoadFileError}}
+              @realmInfo={{this.codeModeRealmInfo}}
+            />
```
we have a component for code mode, why doesn't the URL bar live in that component?

FadhlanR replied:
Because I need to make this component at the same level as submode switcher button. And the submode switcher button is in the container not in the code mode.

habdelra replied:
but why? you can use actions and arguments to shuttle state thru the code mode component

FadhlanR replied:
Do you mean why the submode switcher in the container?

habdelra replied:
no i mean why does having the submode switcher in the container mean that you can't put the URL bar in the code mode component? you haven't given a good reason yet, or justified your statement that it has to be on the same level as the submode switcher

FadhlanR replied:
I did it to make style easier so the submode switcher and code bar URL can be in the same `div`.

habdelra replied:
i think the CSS would be more sophisticated if you moved the URL bar into the code mode component, but that you could totally style it identically to how you have it now using absolute positioning. Personally I think the component hierarchy is more important than the CSS styling. The URL bar only ever appears in code mode so it seems very awkward to have it live outside of the component that is used for code mode.

habdelra replied:
Also, I'm getting concerned about the size of the container.gts file. I think at some point soon we'll need to refactor it and split it apart. it feels like it's losing cohesion and doing too many jobs and thereby attracting other people to add even more to it because it's no longer clear what its specific job is.  

@lukemelia might be a good idea to make a ticket to refactor the container.gts component...

FadhlanR replied:
Agree 

> Also, I'm getting concerned about the size of the container.gts file. I think at some point soon we'll need to refactor it and split it apart. it feels like it's losing cohesion and doing too many jobs and thereby attracting other people to add ever more to it because it's not longer clear what its specific job is.
> 
> @lukemelia might be a good idea to make a ticket to refactor the container.get component...

Agree, I think we can make sub folders for interact mode and code mode.

habdelra replied:
great idea!

FadhlanR replied:
> i think the CSS would be more sophisticated if you moved the URL bar into the code mode component, but that you could totally style it identically to how you have it now using absolute positioning. Personally I think the component hierarchy is more important than the CSS styling. The URL bar only ever appears in code mode so it seems very awkward to have it live outside of the component that is used for code mode.

Updated [152409f](https://github.com/cardstack/boxel/pull/571/commits/152409f06f02aae55163f488c1a8c30d2bb6fd03)

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -586,9 +602,76 @@ export default class OperatorModeContainer extends Component<Signature> {
+    this.loadFileError = null;
+  }
+
+  openFile = maybe(this, (context) => {
```
good!

FadhlanR committed:
Use codePath from operator mode service

FadhlanR committed:
Put code URL in code mode component

habdelra commented on packages/host/app/components/operator-mode/code-mode.gts:
```diff
@@ -1,43 +1,105 @@
+    }
+  });
+
+  openFile = maybe(this, (context) => {
```
i think this is a better place for this--good job

lukemelia commented:
@FadhlanR I don't see anywhere that we are passing the current file or URL into this component, or binding it to a service property. How is this component going to update its URL properly when a file is opened via a method other than typing into the URL bar?
